src/verifast.ml

```diff
@@ -669,19 +726,21 @@ module VerifyProgram(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
           let produce_object envTp =
             if pure then static_error l "Cannot declare a variable of this type in a ghost context." None;
             begin let Some block = !blockPtr in if not (List.mem x !block) then block := x::!block end;
+            let addr_name = x ^ "_addr" in 
+            let addr = get_unique_var_symb_non_ghost addr_name (PtrType Void) in
             let init =
               match e with
-                None -> Unspecified
+              | None -> Unspecified
               | Some e -> Expr (check_c_initializer (pn,ilist) tparams tenv e t)
             in
-            let addr = get_unique_var_symb_non_ghost (x ^ "_addr") (PtrType Void) in
-            produce_c_object l real_unit addr t eval_h init true true h env $. fun h env ->
+            let verify_ctor_call = verify_ctor_call (pn, ilist) leminfo funcmap predinstmap sizemap tenv ghostenv h env addr (Some addr_name) in
+            produce_cxx_object l real_unit addr t eval_h verify_ctor_call init true true h env @@ fun h env ->
             iter h ((x, envTp)::tenv) ghostenv ((x, addr)::env) xs
           in
           match t with
             StaticArrayType (elemTp, elemCount) ->
             produce_object t
-          | StructType sn when !address_taken || (let (_, fds, _, _) = List.assoc sn structmap in match fds with Some fds -> List.exists (fun (_, (_, gh, _, _, _)) -> gh = Ast.Ghost) fds | _ -> true) ->
+          | StructType sn when !address_taken || (language = CLang && dialect = Some Cxx) || (let (_, fds, _, _) = List.assoc sn structmap in match fds with Some fds -> List.exists (fun (_, (_, gh, _, _, _)) -> gh = Ast.Ghost) fds | _ -> true) ->

```

Comment:
It would be better if we could support structs-as-values in C++ as well. Instead of checking for C++, check for the presence of non-C members (i.e. constructors, destructors, or methods)?

Reply:
See https://github.com/verifast/verifast/commit/bf7dc7f662a4c6340cf0fdf9e211334265216e1d

Reply:
The main use case would be passing structs by value as function arguments or return values.

Reply:
In general, it would be good if there were no "regressions": every C program would continue to verify when interpreted as C++.

Reply:
> It would be better if we could support structs-as-values in C++ as well. Instead of checking for C++, check for the presence of non-C members (i.e. constructors, destructors, or methods)?

Currently, when no constructor/destructor is found, an error is given that they should be defined explicitly (because we do not generate default constructors/destructors yet). Checking for the presence of a constructor would also be hard because its mangled name would be required to query its existence.

We cannot check for the existence for methods, because they are translated to functions, i.e., there is no way to easily check if a function would be a member function of an object.

Another possibility would be to add a flag in structs/unions. Clang allows to check if a record is a [POD type](https://en.cppreference.com/w/cpp/named_req/PODType).  That would however probably bloat the code.

Reply:
OK; let's leave it like this for now, then.